compiler/qsc_rca/src/lib.rs

```diff
@@ -0,0 +1,846 @@
+        const CallToUnresolvedCallee = 1 << 16;
+        /// Forward branching on dynamic value.
+        const ForwardBranchingOnDynamicValue = 1 << 17;
+        /// Qubit allocation that happens within a dynamic scope.
+        const DynamicQubitAllocation = 1 << 18;
+        /// Result allocation that happens within a dynamic scope.
+        const DynamicResultAllocation = 1 << 19;
+        /// Use of a dynamic index to access or update an array.
+        const UseOfDynamicIndex = 1 << 20;
+        /// Use of a closure.
+        const UseOfClosure = 1 << 21;
+    }
+}
+
+impl RuntimeFeatureFlags {
+    /// Determines the runtime features that contribute to the provided runtime capabilities.
+    #[must_use]
+    pub fn contributing_features(&self, runtime_capabilities: RuntimeCapabilityFlags) -> Self {
+        let mut contributing_features = Self::empty();
+        for feature in self.iter() {
+            if feature
+                .runtime_capabilities()
+                .intersects(runtime_capabilities)
+            {
+                contributing_features |= feature;
+            }
+        }
+
+        contributing_features
+    }
+
+    /// Maps program contructs to runtime capabilities.
+    #[must_use]
+    pub fn runtime_capabilities(&self) -> RuntimeCapabilityFlags {
+        let mut runtume_capabilities = RuntimeCapabilityFlags::empty();
+        if self.contains(RuntimeFeatureFlags::UseOfDynamicBool) {
+            runtume_capabilities |= RuntimeCapabilityFlags::ForwardBranching;
+        }
+        if self.contains(RuntimeFeatureFlags::UseOfDynamicInt) {
+            runtume_capabilities |= RuntimeCapabilityFlags::IntegerComputations;
+        }
+        if self.contains(RuntimeFeatureFlags::UseOfDynamicPauli) {
+            runtume_capabilities |= RuntimeCapabilityFlags::IntegerComputations;
+        }
+        if self.contains(RuntimeFeatureFlags::UseOfDynamicRange) {
+            runtume_capabilities |= RuntimeCapabilityFlags::IntegerComputations;
+        }
+        if self.contains(RuntimeFeatureFlags::UseOfDynamicDouble) {
+            runtume_capabilities |= RuntimeCapabilityFlags::FloatingPointComputations;
+        }
+        if self.contains(RuntimeFeatureFlags::UseOfDynamicQubit) {
+            runtume_capabilities |= RuntimeCapabilityFlags::HigherLevelConstructs;
+        }
+        if self.contains(RuntimeFeatureFlags::UseOfDynamicBigInt) {
+            runtume_capabilities |= RuntimeCapabilityFlags::HigherLevelConstructs;
+        }
+        if self.contains(RuntimeFeatureFlags::UseOfDynamicString) {
+            runtume_capabilities |= RuntimeCapabilityFlags::HigherLevelConstructs;
+        }
+        if self.contains(RuntimeFeatureFlags::UseOfDynamicallySizedArray) {
+            runtume_capabilities |= RuntimeCapabilityFlags::HigherLevelConstructs;
+        }
+        if self.contains(RuntimeFeatureFlags::UseOfDynamicUdt) {
+            runtume_capabilities |= RuntimeCapabilityFlags::all();
```

Comment:
I wonder if using `all()` here might be confusing... it's not that we know it's using every capability but rather we can no longer verify what capabilities it uses because it's too complex. My reading of "Higher level constructs" was that this was meant to be the catch all for "things we can't handle yet." I think it will work either way, just wanted to confirm the intention.

Reply:
I agree, using `all()` here might be confusing and I think it was me being overly cautious before more pieces were in place. I'll update this function to replace `all()` with `HigherLevelConstructs`.

Reply:
Updated.